From fb84e13717eb821ecc5395543a2539a3f13cf141 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sat, 23 Nov 2024 01:26:07 +0100
Subject: [PATCH] [Dogecash] Fix "Illegal use of SIGHASH_FORKID" in wallet
 (#79)

When the user provides no sighash, it defaulted to using `withForkId` when parsing the sighash. Now it toggles based on settings.

diff --git a/src/core_read.cpp b/src/core_read.cpp
index d079267a010d248579a132d8b09da3fdc1f216c9..f41e7f55e4f45b667ace21c2c3e2326b78779730 100644
--- a/src/core_read.cpp
+++ b/src/core_read.cpp
@@ -259,7 +259,10 @@ std::vector<uint8_t> ParseHexUV(const UniValue &v, const std::string &strName) {
 }
 
 SigHashType ParseSighashString(const UniValue &sighash) {
-    SigHashType sigHashType = SigHashType().withForkId();
+    SigHashType sigHashType =
+        GetSignScriptFlags() & SCRIPT_ENABLE_SIGHASH_FORKID
+            ? SigHashType().withForkId()
+            : SigHashType();
     if (!sighash.isNull()) {
         static const std::map<std::string, int> map_sighash_values = {
             {"ALL", SIGHASH_ALL},
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index e3cca5eaef2d4fe8ebd517d5536dd0fa4b3edf0a..81fb31cf276f4c8f15bf570c8cc4a0850e623a6f 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -2104,7 +2104,11 @@ bool CWallet::SignTransaction(CMutableTransaction &tx) const {
                  wtx.IsCoinBase());
     }
     std::map<int, std::string> input_errors;
-    return SignTransaction(tx, coins, SigHashType().withForkId(), input_errors);
+    return SignTransaction(tx, coins,
+                           GetSignScriptFlags() & SCRIPT_ENABLE_SIGHASH_FORKID
+                               ? SigHashType()
+                               : SigHashType().withForkId(),
+                           input_errors);
 }
 
 bool CWallet::SignTransaction(CMutableTransaction &tx,
diff --git a/test/functional/feature_legacy_signrawtransaction.py b/test/functional/feature_legacy_signrawtransaction.py
index d7d495c4fb3e77a8a8112b757e4988495c7d4f90..7da9f139837703228a61bb16a66948dfe872167b 100644
--- a/test/functional/feature_legacy_signrawtransaction.py
+++ b/test/functional/feature_legacy_signrawtransaction.py
@@ -4,6 +4,7 @@
 """Test we sign valid Dogecoin legacy signatures for transactions."""
 
 from test_framework.address import ADDRESS_ECREG_UNSPENDABLE
+from test_framework.hash import hash160
 from test_framework.key import ECKey
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
 from test_framework.script import (
@@ -15,7 +16,7 @@ from test_framework.script import (
     hash160,
 )
 from test_framework.test_framework import BitcoinTestFramework
-from test_framework.util import assert_raises_rpc_error
+from test_framework.util import assert_equal, assert_raises_rpc_error
 from test_framework.wallet_util import bytes_to_wif
 
 
@@ -32,6 +33,11 @@ class LegacyScriptSignrawtransactionTest(BitcoinTestFramework):
         node = self.nodes[0]
         nonlegacy_node = self.nodes[1]
 
+        def check_signed_result(node, result):
+            assert result["complete"], f"Signing failed: {result}"
+            accept_result = node.testmempoolaccept([result["hex"]])[0]
+            assert_equal(accept_result["allowed"], True)
+
         coinblockhash1 = self.generate(node, 1)[0]
         cointxid1 = node.getblock(coinblockhash1)["tx"][0]
 
@@ -67,7 +73,7 @@ class LegacyScriptSignrawtransactionTest(BitcoinTestFramework):
         ]
         for sighash in legacy_sighashes:
             result = node.signrawtransactionwithwallet(tx.serialize().hex(), [], sighash)
-            assert result["complete"]
+            check_signed_result(node, result)
             assert_raises_rpc_error(
                 -8,
                 "Signature must use SIGHASH_FORKID",
@@ -97,11 +103,11 @@ class LegacyScriptSignrawtransactionTest(BitcoinTestFramework):
             result = nonlegacy_node.signrawtransactionwithwallet(
                 nonlegacy_tx.serialize().hex(), [], sighash
             )
-            assert result["complete"]
+            check_signed_result(nonlegacy_node, result)
 
         # Legacy tx works
         signed_tx = node.signrawtransactionwithwallet(tx.serialize().hex(), [], "ALL")
-        assert signed_tx["complete"]
+        check_signed_result(node, signed_tx)
         txid = node.sendrawtransaction(signed_tx["hex"])
         assert_raises_rpc_error(
             -26,
@@ -113,7 +119,7 @@ class LegacyScriptSignrawtransactionTest(BitcoinTestFramework):
         signed_tx = nonlegacy_node.signrawtransactionwithwallet(
             nonlegacy_tx.serialize().hex(), [], "ALL|FORKID"
         )
-        assert signed_tx["complete"]
+        check_signed_result(nonlegacy_node, signed_tx)
         nonlegacy_txid = nonlegacy_node.sendrawtransaction(signed_tx["hex"])
         # Note: NULLFAIL is not mandatory on legacy, so we just get a FALSE from OP_CHECKSIG
         assert_raises_rpc_error(
@@ -131,7 +137,8 @@ class LegacyScriptSignrawtransactionTest(BitcoinTestFramework):
         nonlegacy_tx2.vin = [CTxIn(COutPoint(int(nonlegacy_txid, 16), 0))]
         nonlegacy_tx2.vout = [CTxOut(coinvalue - 2000, p2pkh)]
 
-        tx2_keys = [bytes_to_wif(key.get_bytes())]
+        tx2_key = bytes_to_wif(key.get_bytes())
+        tx2_keys = [tx2_key]
         tx2_outpoints = [
             {
                 "txid": txid,
@@ -148,11 +155,45 @@ class LegacyScriptSignrawtransactionTest(BitcoinTestFramework):
                 "amount": tx.vout[0].nValue / 100,
             }
         ]
+
+        # Import keys to wallet
+        node.importprivkey(tx2_key)
+        nonlegacy_node.importprivkey(tx2_key)
+
+        # Test sigs without args (defaults to ALL)
+        result = node.signrawtransactionwithkey(
+            tx2.serialize().hex(), tx2_keys, tx2_outpoints
+        )
+        check_signed_result(node, result)
+
+        result = node.signrawtransactionwithwallet(
+            tx2.serialize().hex(), tx2_outpoints
+        )
+        check_signed_result(node, result)
+
+        result = nonlegacy_node.signrawtransactionwithkey(
+            nonlegacy_tx2.serialize().hex(), tx2_keys, nonlegacy_tx2_outpoints
+        )
+        check_signed_result(nonlegacy_node, result)
+
+        result = nonlegacy_node.signrawtransactionwithwallet(
+            nonlegacy_tx2.serialize().hex(), nonlegacy_tx2_outpoints
+        )
+        check_signed_result(nonlegacy_node, result)
+
         for sighash in legacy_sighashes:
+            # Test signing with key
             result = node.signrawtransactionwithkey(
                 tx2.serialize().hex(), tx2_keys, tx2_outpoints, sighash
             )
-            assert result["complete"]
+            check_signed_result(node, result)
+
+            # Test signing with wallet
+            result = node.signrawtransactionwithwallet(
+                tx2.serialize().hex(), tx2_outpoints, sighash
+            )
+            check_signed_result(node, result)
+
             assert_raises_rpc_error(
                 -8,
                 "Signature must use SIGHASH_FORKID",
@@ -162,6 +203,16 @@ class LegacyScriptSignrawtransactionTest(BitcoinTestFramework):
                 nonlegacy_tx2_outpoints,
                 sighash,
             )
+
+            assert_raises_rpc_error(
+                -8,
+                "Signature must use SIGHASH_FORKID",
+                nonlegacy_node.signrawtransactionwithwallet,
+                nonlegacy_tx2.serialize().hex(),
+                nonlegacy_tx2_outpoints,
+                sighash,
+            )
+
         for sighash in bip143_sighashes:
             assert_raises_rpc_error(
                 -8,
@@ -172,13 +223,30 @@ class LegacyScriptSignrawtransactionTest(BitcoinTestFramework):
                 tx2_outpoints,
                 sighash,
             )
+
+            assert_raises_rpc_error(
+                -8,
+                "Illegal use of SIGHASH_FORKID",
+                node.signrawtransactionwithwallet,
+                tx2.serialize().hex(),
+                tx2_outpoints,
+                sighash,
+            )
+
             result = nonlegacy_node.signrawtransactionwithkey(
                 nonlegacy_tx2.serialize().hex(),
                 tx2_keys,
                 nonlegacy_tx2_outpoints,
                 sighash,
             )
-            assert result["complete"]
+            check_signed_result(nonlegacy_node, result)
+
+            result = nonlegacy_node.signrawtransactionwithwallet(
+                nonlegacy_tx2.serialize().hex(),
+                nonlegacy_tx2_outpoints,
+                sighash,
+            )
+            check_signed_result(nonlegacy_node, result)
 
 
 if __name__ == "__main__":
